spec/src/main/asciidoc/property-expressions.asciidoc

```diff
@@ -0,0 +1,79 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// Contributors:
+// Roberto Cortez
+
+[[property-expressions]]
+== Property Expressions
+
+The value of a configuration property may contain an expression corresponding to another configuration property. An
+expression string is a mix of plain strings and expression segments, which are wrapped by the sequence `${ ... }`.
+
+Consider the following configuration properties file:
+
+[source,properties]
+----
+server.url=http://${server.host}/endpoint
+server.host=example.org
+----
+
+When looking up the `server.url` property, the value will be resolved and expanded to `http://example.org/endpoint`.
+All MicroProfile Config rules still apply. The `Config` is able to resolve expressions from different ConfigSources.
+
+Additionally, it is also possible to use the following syntax for property expressions:
+
+* `${expression:value}` - Provides a default value after the `:` if the expression doesn't find a value.
+* `${my.prop${compose}}` - Composed expressions. Inner expressions are resolved first.
+* `${my.prop}${my.prop}` - Multiple expressions.
+
+Consider the following configuration properties file:
+
+[source,properties]
+----
+server.url=http://${server.host:example.org}:${server.port}/${server.endpoint}
+server.port=8080
+server.endpoint=${server.endpoint.path.foo${server.endpoint.path.bar}}
+server.endpoint.path.foo=/foo
+server.endpoint.path.bar=/bar
+----
+
+The property `server.url` is expanded to `http://example.org:8080/foo/bar`.
+
+If an expression cannot be expanded and does not have a default value, a `NoSuchElementException` is thrown. In the
+Optional case, an empty Optional will be returned.
+
+The number of recursion lookups is not infinite, but a limited number for composed expressions. Implementations are
+encouraged to limit the number to `5`, but they can use higher limits if they wish to. When the number of allowed
+lookups exceeds the limit, a `IllegalArgumentException` is thrown.
+
+Property expressions applies to all the methods in `Config` that performs resolution of a configuration property,
+including `getValue`, `getValues`, `getConfigValue`, `getValues`, `getOptionalValue`, `getOptionalValues`
+and `getConfigProperties`. The methods `getValue` and `getProperties` in `ConfigSource`, may support property
+expressions as well, but it is not required by the specification.
+
+The escape character is "\".
+
+=== Backwards Compatibility
```

Comment:
I checked and we already use `${...}` syntax for a particular source. We cannot change that syntax as it  originates from other concepts that must stay as they are. 

I think the way to integrate is simply to have this source interpret the syntax with its own mechanism while all other sources use the MP mechanism. From outside of the source it will look like there was no expression used as it is already resolved when the values exists the source.

Maybe there should be something here in the spec saying that special sources might do something else when encountering this syntax but the sources named in the spec will honour it and in general sources try to follow the spec.

Reply:
Doesn't this piece covers it?

> Property expressions applies to all the methods in `Config` that performs resolution of a configuration property, including `getValue`, `getValues`, `getConfigValue`, `getValues`, `getOptionalValue`, `getOptionalValues` and `getConfigProperties`. The methods `getValue` and `getProperties` in `ConfigSource`, may support property expressions as well, but it is not required by the specification.

Reply:
To me this is about the spec mechanism being implemented on the source level or not. What I mean is that there is a complete different mechanism in place that could be confused with MP Config expressions because of the same syntax.

Reply:
@jbee you are asking for a config source level disablement. I assume the variable expansion is done at last. By then, it is difficult to work out where the config value is from.

Reply:
I am not asking for a feature, just that the spec acknowledges that there might be sources where the same syntax has different effect.

Reply:
@jbee, please feel free to propose what you think it makes sense.

Reply:
@jbee I understood your scenario for skipping the expansion for some special config sources. Yep, I am also ok for adding some relaxation sentence.